File: ch09/go/examples.go
package ch09

import (
	"fmt"
	"slices"
)

type Stack struct {
	data []any
}

func (s *Stack) Push(element any) {
	s.data = append(s.data, element)
}

func (s *Stack) Pop() any {
	var element any

	if len(s.data) > 0 {
		element = s.data[len(s.data)-1]
		s.data = s.data[:len(s.data)-1]
	}

	return element
}

func (s *Stack) Read() any {
	if len(s.data) > 0 {
		return s.data[len(s.data)-1]
	} else {
		return nil
	}
}

type Linter struct {
	stack Stack
}

func (l *Linter) Lint(text string) string {
	for _, char := range text {
		if l.isOpeningBrace(char) {
			l.stack.Push(char)
		} else if l.isClosingBrace(char) {
			poppedOpeningBrace := l.stack.Pop()

			if poppedOpeningBrace == nil {
				return string(char) + " doesn't have an opening brace"
			}

			if l.isNotMatch(poppedOpeningBrace.(rune), char) {
				return string(char) + " has mismatched opening brace"
			}
		}
	}

	if l.stack.Read() != nil {
		return string(l.stack.Read().(rune)) + " does not have a closing brace"
	}

	return "No errors"
}

func (l *Linter) isOpeningBrace(char rune) bool {
	openingBraces := []rune{'(', '[', '{'}

	return slices.Contains(openingBraces, char)
}

func (l *Linter) isClosingBrace(char rune) bool {
	closingBraces := []rune{')', ']', '}'}

	return slices.Contains(closingBraces, char)
}

func (l *Linter) isNotMatch(openingBrace rune, closingBrace rune) bool {
	if (openingBrace == '(' && closingBrace == ')') ||
		(openingBrace == '[' && closingBrace == ']') ||
		(openingBrace == '{' && closingBrace == '}') {
		return false
	}

	return true
}

type Queue struct {
	data []any
}

func (q *Queue) Enqueue(element any) {
	q.data = append(q.data, element)
}

func (q *Queue) Dequeue() any {
	var element any

	if len(q.data) > 0 {
		element = q.data[0]
		q.data = q.data[1:]
	}

	return element
}

func (q *Queue) Read() any {
	if len(q.data) > 0 {
		return q.data[0]
	} else {
		return nil
	}
}

type PrintManager struct {
	queue Queue
}

func (pm *PrintManager) QueuePrintJob(document any) {
	pm.queue.Enqueue(document)
}

func (pm *PrintManager) Run() {
	for {
		if pm.queue.Read() != nil {
			pm.print(pm.queue.Dequeue())
		} else {
			break
		}
	}
}

func (pm *PrintManager) print(document any) {
	fmt.Println("Document:", document)
}

File: ch09/python/examples.py
"""
Chapter 09: Examples
"""


class Stack:
    def __init__(self):
        self.data = []
    
    def push(self, element):
        self.data.append(element)

    def pop(self):
        if len(self.data) > 0:
            return self.data.pop()
        else:
            return None

    def read(self):
        if len(self.data) > 0:
            return self.data[-1]
        else:
            return None


class Linter:
    def __init__(self):
        self.stack = Stack()

    def lint(self, text):
        for char in text:
            if self._is_opening_brace(char):
                self.stack.push(char)
            elif self._is_closing_brace(char):
                popped_opening_brace = self.stack.pop()

                if not popped_opening_brace:
                    return f"{char} doesn't have an opening brace"
                
                if self._is_not_match(popped_opening_brace, char):
                    return f"{char} has mismatched opening brace"
                
        if self.stack.read():
            return f"{self.stack.read()} does not have a closing brace"

    def _is_opening_brace(self, char):
        return char in ["(", "[", "{"]
    
    def _is_closing_brace(self, char):
        return char in [")", "]", "}"]
    
    def _is_not_match(self, opening_brace, closing_brace):
        if (
            (opening_brace == "(" and closing_brace == ")")
            or (opening_brace == "[" and closing_brace == "]")
            or (opening_brace == "{" and closing_brace == "}")
        ):
            return False
        
        return True
    

class Queue:
    def __init__(self):
        self.data = []

    def enqueue(self, element):
        self.data.append(element)

    def dequeue(self):
        if len(self.data) > 0:
            return self.data.pop(0)
        else:
            return None
    
    def read(self):
        if len(self.data) > 0:
            return self.data[0]
        else:
            return None
        

class PrintManager:
    def __init__(self):
        self.queue = Queue()

    def queue_print_job(self, document):
        self.queue.enqueue(document)

    def run(self):
        while self.queue.read():
            self._print(self.queue.dequeue())

    def _print(self, document):
        print(f"Document: {document}")


print("******** Stack ********")
stack = Stack()
assert stack.pop() is None
assert stack.read() is None

stack.push(1)
stack.push(2)
stack.push(3)
assert stack.read() == 3
assert stack.pop() == 3
assert stack.read() == 2


print("******** Linter ********")
assert Linter().lint("(a)") is None
assert Linter().lint("(a[b]{c})") is None
assert Linter().lint("(a") == "( does not have a closing brace"
assert Linter().lint("(a]") == "] has mismatched opening brace"
assert Linter().lint("a)") == ") doesn't have an opening brace"


print("******** Queue ********")
queue = Queue()
assert queue.dequeue() is None
assert queue.read() is None

queue.enqueue("first.pdf")
queue.enqueue("second.pdf")
queue.enqueue("third.pdf")
assert queue.read() == "first.pdf"
assert queue.dequeue() == "first.pdf"
assert queue.read() == "second.pdf"


print("******** Print Manager ********")
print_manager = PrintManager()
print_manager.queue_print_job("first.pdf")
print_manager.queue_print_job("second.pdf")
print_manager.queue_print_job("third.pdf")
print_manager.run()